cli/EAP7-924_Coloured_output_for_CLI.adoc

```diff
@@ -0,0 +1,83 @@
+= EAP7-924 Coloured output for CLI
+:author:            Ingo Weiss
+:email:             ingo@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+:issue-base-url:    https://issues.jboss.org/browse/
+
+== Overview
+
+The intent of this feature is to provide a better user experience by means of colouring important parts of the CLI workflow, this includes error messages, warning messages, successful outputs and better visual feedback when using batch and flow control mechanisms if/for/while.
+
+== Issue Metadata
+
+=== Issue:
+
+* {issue-base-url}EAP7-924[EAP7-924]
+
+=== Related Issues:
+
+* {issue-base-url}WFCORE-3450
+* {issue-base-url}WFCORE-3577
+
+=== Dev Contacts:
+
+* mailto:ingo@redhat.com[Ingo Weiss]
+* mailto:jdenise@redhat.com[Jean-François Denise]
+
+=== QE Contacts:
+
+* TBD
+
+=== Affected Projects or Components:
+
+* WildFly Core CLI
+
+
+== Requirements
+
+=== Hard requirements
+
+* Ability to enable or disable colour output via CLI argument
+* Ability to enable or disable colour output using `jboss-cli.xml`
+* Ability to configure the colours for each allowed type: `warning`, `error`, `success`, `batch`, `required`
+* Mark required attributes in bold and with a different colour from the default
+* Colour the output of DMR output based on its outcome
+* Add colours to CLI prompt
+
+=== Soft requirements
+
+* Have a different prompt for each workflow step
+
+== Test Plan
+
+The following "matrix" should be tested on each WildFly/EAP supported operating system:
+
+
+[width="75%",options="header,footer"]
+|===
+| Terminal colours? | CLI colour
+| YES | ON
+| YES | OFF
+| NO  | ON
+| NO  | OFF
+|===
+
+== Design Notes
+
+The current prototype assumes the following defaults:
+
+[width="75%",options="header,footer"]
+|===
+| Type | Colour | Intensity
+| Error | Red | Bright
+| Warning | Yellow | Normal
+| Success | Green | Normal
+| Required | Magenta | Bright
+| Batch | Green | Bright
+|===
+
+These defaults might need to be adjusted depending on the target platform. For example, Windows doesn't render the colour yellow with normal intensity. The UX team might be of great help in setting these defaults.
```

Comment:
I like the idea to expand to a full 256 colours, but right now we are tied to what Aesh Readline can provide.

The ideal of a standard set of system properties is interesting but I can't see how we could combine logging with CLI, for example, but I'm all in on the idea :-)

Reply:
It's not about combining as much as it is about having a standard blob in our shell scripts to detect the availability of color and information about it, which can be passed in to our Java applications which in turn can then look into a single place for it.

If the console supports 256 colors or truecolor, then it also supports 8 and 16 color modes.

Reply:
@iweiss I think this deserves mention in the design document. Can you please elaborate how you plan to detect color capabilities of terminals?

Reply:
The terminal capabilities are handled by AESH Readline already and, on my limited tests, it works without issues.

Reply:
aesh has some logic (thanks to infocmp) to detect the maximum number of colors but actually this information is not in use (both by CLI and aesh). ANSI codes are printed, if the terminal support colors, it will print colors, otherwise will use default colors.

I don't see a need for more than a limited set of 8 colors and wouldn't want to expose to the user some complex configuration to tune colors/rendering. This set of colors is "abstract" and, if required, is adjusted internally according to the platform (eg: if windows, Yellow is bright, seems that this is the only case we ran into). So I am not sure the CLI would benefit from 256 colors. 
Adding platform specific logic in scripts shows to be difficult to adjust, I think that some strong requirements should push us to invest in this. The CLI doesn't seem to be one.

Reply:
@dmlloyd @jfdenise Are there any more considerations or should we try and add this to WF13?

Reply:
I am ok with the proposal, I don't have concerns.